cmd/kube-controller-manager/app/certificates.go

```diff
@@ -222,35 +239,37 @@ func startRootCACertificatePublisherController(ctx context.Context, controllerCo
 		rootCA,
 	)
 	if err != nil {
-		return nil, true, fmt.Errorf("error creating root CA certificate publisher: %v", err)
+		return fmt.Errorf("error creating root CA certificate publisher: %v", err)
 	}
-	go sac.Run(ctx, 1)
-	return nil, true, nil
+	sac.Run(ctx, 1)
+	return nil
 }
 
 func newKubeAPIServerSignerClusterTrustBundledPublisherDescriptor() *ControllerDescriptor {
 	return &ControllerDescriptor{
-		name:                 names.KubeAPIServerClusterTrustBundlePublisherController,
-		initFunc:             newKubeAPIServerSignerClusterTrustBundledPublisherController,
+		name: names.KubeAPIServerClusterTrustBundlePublisherController,
+		initFunc: func(ctx context.Context, controllerContext ControllerContext, controllerName string) (Runnable, error) {
+			rootCA, err := getKubeAPIServerCAFileContents(controllerContext)
+			if err != nil {
+				return nil, err
+			}
+			if len(rootCA) == 0 {
+				return nil, nil
+			}
+			return runnableFunc(func(ctx context.Context) error {
+				return startKubeAPIServerSignerClusterTrustBundledPublisherController(ctx, controllerContext, controllerName, rootCA)
```

Comment:
Generally it would be more correct to separate the error from init and error from start, but in this current proposal many init errors are just returned form the start function as otherwise there would be much more boilerplate code. But if we agree on an approach, I don't mind splitting it properly in all init functions.

Reply:
This seems to be a pre-existing problem, so I'm personally fine with keeping it like this.

Reply:
Are non-init errors currently possible? Seems like most controllers invoke `Run()` in a goroutine without errors being returned from there. One thing I'm a bit worried now is whether a single runtime `Start()` could cause KCM to shutdown since we use `errgroup.Wait()` that will cancel the context if any controllers return an error.

Reply:
Yeah, currently all init functions are run sequentially and any error causes KCM to terminate. Just in the middle of all this synchronous code, there are some go Run() calls. I am actually going to refactor this change so that it's as close to the current behavior as possible, it's anyway needed for properly setting up the informers, I think.

Reply:
Actually no `Run` function ever returns an error, true, so I will consider changing the interface.